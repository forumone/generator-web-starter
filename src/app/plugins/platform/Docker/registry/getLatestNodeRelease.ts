import got from 'got';
import semver from 'semver';

interface Release {
  readonly version: string;
  readonly date: string;
  readonly files: ReadonlyArray<string>;
  readonly npm?: string;
  readonly v8: string;
  readonly uv?: string;
  readonly openssl?: string;
  readonly modules?: string;
  readonly lts: string | false;
  readonly security: boolean;
}

export interface Dist {
  readonly version: string;
  readonly checksum: string;
}

async function getLatestRelease(): Promise<Release> {
  const response = await got('https://nodejs.org/dist/index.json', {
    json: true,
  });
  const versions: Release[] = response.body;

  // Compare b <=> a for descending sort
  versions.sort((a, b) => semver.compare(b.version, a.version));

  return versions.find(version => version.lts !== false) || versions[0];
}

async function getReleaseChecksum(version: string): Promise<string> {
  const response = await got(
    `https://nodejs.org/dist/${version}/SHASUMS256.txt`,
  );

  const checksums = response.body;

  // NB. SHASUMS256.txt is formatted like so:
  // 9a16909157e68d4e409a73b008994ed05b4b6bc952b65ffa7fbc5abb973d31e9  node-v12.4.0-linux-x64.tar.gz
  // This means that splitting on lines and then whitespace is sufficient to "parse" the string
  // into a checksum/file pair.
  const parsed = checksums
    .split('\n')
    .map(line => line.split(/\s+/) as [string, string]);

  const releaseFile = `node-${version}-linux-x64.tar.gz`;
  const matching = parsed.find(pair => pair[1] === releaseFile);
  if (!matching) {
    throw new Error(
      `Unable to find checksum for Node.js release ${releaseFile}`,
    );
  }

  return matching[0];
}

async function getLatestNodeRelease(): Promise<Dist> {
  const release = await getLatestRelease();
  const checksum = await getReleaseChecksum(release.version);

  return {
    checksum,
    version: release.version,
  };
}

export default getLatestNodeRelease;
