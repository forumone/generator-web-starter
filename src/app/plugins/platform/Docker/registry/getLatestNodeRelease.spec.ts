import getLatestNodeRelease from './getLatestNodeRelease';

let got!: jest.Mock;

beforeEach(() => {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  got = require('got');
  got.mockClear();
});

const listingWithLts = [
  { version: 'v1.2.3', lts: false },
  { version: 'v4.1.0', lts: 'uranium' },
  { version: 'v2.3.1', lts: false },
];

const listingWithoutLts = [
  { version: 'v1.2.3', lts: false },
  { version: 'v2.3.1', lts: false },
];

const checksumsForLts = `
11223344 node-v4.1.0-linux-x64.tar.xz
55667788 node-v4.1.0-darwin-x64.tar.gz
AABBCCDD node-v4.1.0-linux-x64.tar.gz
`;

const checksumsForNotLts = `
11223344 node-v1.2.3-linux-x64.tar.gz
55667788 node-v2.3.1-linux-x64.tar.gz
`;

const badChecksums = '';

test.each`
  type         | version     | checksum      | listing              | checksums
  ${'LTS'}     | ${'v4.1.0'} | ${'AABBCCDD'} | ${listingWithLts}    | ${checksumsForLts}
  ${'not LTS'} | ${'v2.3.1'} | ${'55667788'} | ${listingWithoutLts} | ${checksumsForNotLts}
`(
  'Finds $type release $version',
  async ({ version, checksum, listing, checksums }) => {
    got
      .mockResolvedValueOnce({ body: listing })
      .mockResolvedValue({ body: checksums });

    const latestNodeRelease = await getLatestNodeRelease();

    expect(got).toBeCalledTimes(2);
    expect(got).toHaveBeenNthCalledWith(
      1,
      'https://nodejs.org/dist/index.json',
      { json: true },
    );
    expect(latestNodeRelease.version).toEqual(version);
    expect(latestNodeRelease.checksum).toEqual(checksum);
  },
);

test.each`
  type         | listing
  ${'LTS'}     | ${listingWithLts}
  ${'not LTS'} | ${listingWithoutLts}
`('Fails to parse checksums for $type releases', async ({ listing }) => {
  got
    .mockResolvedValueOnce({ body: listing })
    .mockResolvedValue({ body: badChecksums });

  await expect(getLatestNodeRelease()).rejects.toThrowErrorMatchingSnapshot();
});
