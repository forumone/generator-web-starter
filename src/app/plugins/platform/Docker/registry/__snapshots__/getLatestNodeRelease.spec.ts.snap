// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fails to parse checksums for LTS releases 1`] = `"Unable to find checksum for Node.js release node-v4.1.0-linux-x64.tar.gz"`;

exports[`Fails to parse checksums for not LTS releases 1`] = `"Unable to find checksum for Node.js release node-v2.3.1-linux-x64.tar.gz"`;
